File: src/main/java/com/github/frcsty/litebansdiscord/discord/command/HistoryCommand.java
package com.github.frcsty.litebansdiscord.discord.command;

import com.github.frcsty.litebansdiscord.DiscordPlugin;
import com.github.frcsty.litebansdiscord.discord.util.InformationHolder;
import com.github.frcsty.litebansdiscord.discord.util.Utilities;
import me.mattstudios.mfjda.annotations.Command;
import me.mattstudios.mfjda.annotations.Default;
import me.mattstudios.mfjda.annotations.Prefix;
import me.mattstudios.mfjda.base.CommandBase;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageChannel;
import net.dv8tion.jda.api.entities.User;
import org.apache.commons.lang.StringUtils;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;

import java.awt.*;
import java.util.Date;
import java.util.List;

@Prefix("-")
@Command("history")
public final class HistoryCommand extends CommandBase {

    private final DiscordPlugin plugin;

    public HistoryCommand(final DiscordPlugin plugin) {
        this.plugin = plugin;
    }

    @Default
    public void historyCommand() {
        System.out.println("Ran history command");
        final Message message = getMessage();
        final MessageChannel channel = message.getChannel();
        final User user = message.getAuthor();

        if (Utilities.isNotUser(user, message)) return;
        if (message.getMember() == null) return;
        if (Utilities.hasMissingPermission(message.getMember(), plugin.getConfig().getString("settings.requiredRoleId"))) {
            channel.sendMessage("You do not have the required permission for this!").queue();
            return;
        }

        final String[] args = message.getContentRaw().split(" ");
        if (args.length == 1) {
            channel.sendMessage("You did not specify a user to check!").queue();
            return;
        }

        final OfflinePlayer player = Bukkit.getOfflinePlayer(args[1]);
        final List<InformationHolder> holders = Utilities.getUserBans(player);
        final int fromPosition = args.length < 3 ? 0 : Integer.valueOf(args[2]);
        final int toPosition = args.length < 4 ? fromPosition + 5 : Integer.valueOf(args[3]);

        if (holders.size() == 0) {
            channel.sendMessage("User " + player.getName() + " has no history.").queue();
            return;
        }

        channel.sendMessage(getFormattedEmbed(holders, player, fromPosition, toPosition).build()).queue();
        holders.clear();
    }

    private EmbedBuilder getFormattedEmbed(final List<InformationHolder> holders, final OfflinePlayer player, final int from, final int to) {
        final EmbedBuilder embedBuilder = new EmbedBuilder().setColor(Color.CYAN);
        final StringBuilder builder = new StringBuilder();
        embedBuilder.setTitle("History for " + player.getName() + " (Limit: " + holders.size() + "):");

        for (int i = from; i < to; i++) {
            if (holders.size() < i + 1) continue;
            final InformationHolder holder = holders.get(i);

            if (holder == null) continue;
            builder.append(getFormattedHistoryInformation(holder, i));
        }

        embedBuilder.setDescription(builder.toString());
        return embedBuilder;
    }

    private StringBuilder getFormattedHistoryInformation(final InformationHolder holder, final int position) {
        final StringBuilder builder = new StringBuilder();

        builder.append("```\n");
        builder.append("[" + (position + 1) + "] Time: " + new Date(holder.getTime()) + "\n");
        builder.append(" - Reason: " + holder.getReason() + "\n");
        builder.append(" - Banned By: " + holder.getPunisher() + "\n");

        if (holder.isActive()) {
            builder.append(" - Expires: " + new Date(holder.getUntil()) + "\n");
        }

        builder.append(" - Active: " + StringUtils.capitalize(String.valueOf(holder.isActive())) + "\n");
        builder.append("```");

        return builder;
    }
}

File: src/main/kotlin/com/github/frcsty/litebansdiscord/discord/command/HistoryCommand.kt
package com.github.frcsty.litebansdiscord.discord.command

import com.github.frcsty.litebansdiscord.DiscordPlugin
import com.github.frcsty.litebansdiscord.discord.util.InformationHolder
import com.github.frcsty.litebansdiscord.discord.util.getOfflineUser
import com.github.frcsty.litebansdiscord.discord.util.getUserBans
import com.github.frcsty.litebansdiscord.discord.util.isNotMember
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import org.apache.commons.lang.StringUtils
import org.bukkit.OfflinePlayer
import java.awt.Color
import java.util.*

class HistoryCommand(private val plugin: DiscordPlugin) : ListenerAdapter() {

    override fun onGuildMessageReceived(event: GuildMessageReceivedEvent) {
        val start = System.currentTimeMillis()
        val message = event.message
        val content = message.contentRaw
        if (content.startsWith("${plugin.prefix}history").not() && !content.startsWith(event.guild.selfMember.asMention)) {
            return
        }

        val channel = event.channel
        val user = event.author

        if (user.isNotMember(message)) {
            return
        }
        /*
        if (message.member.hasMissingPermission(plugin.config.getString("settings.requiredRoleId"))) {
            channel.sendMessage("You do not have the required permission for this!").queue()
            return
        }
        */

        val args = content.split(" ").toTypedArray()
        if (args.size == 1) {
            channel.sendMessage("You did not specify a user to check!").queue()
            return
        }

        val player = getOfflineUser(args[1])
        if (player == null) {
            channel.sendMessage("Specified user does not exist! (User: ${args[1]})").queue()
            return
        }
        val holders = player.getUserBans()
        val fromPosition = if (args.size < 3) 0 else Integer.valueOf(args[2])
        val toPosition = if (args.size < 4) fromPosition + 5 else Integer.valueOf(args[3])
        if (holders.isEmpty()) {
            channel.sendMessage("User ${player.name} has no history.").queue()
            return
        }

        channel.sendMessage(getFormattedEmbed(holders, player, fromPosition, toPosition, start).build()).queue()

    }

    private fun getFormattedEmbed(holders: List<InformationHolder>, player: OfflinePlayer, from: Int, to: Int, start: Long): EmbedBuilder {
        val embedBuilder = EmbedBuilder().setColor(Color.CYAN)
        val builder = StringBuilder()
        embedBuilder.setTitle("History for ${player.name} (Limit: ${holders.size}):")

        for (i in from until to) {
            if (holders.size < i + 1) continue
            val holder = holders[i]
            builder.append(getFormattedHistoryInformation(holder, i))
        }

        embedBuilder.setDescription(builder.toString())
        embedBuilder.setFooter("Executed in ${System.currentTimeMillis() - start}ms", null)
        return embedBuilder
    }

    private fun getFormattedHistoryInformation(holder: InformationHolder, position: Int): StringBuilder {
        val time = Date(holder.time)
        val expiry = Date(holder.until)
        val active = StringUtils.capitalize(holder.isActive.toString())

        val builder = StringBuilder()
        builder.append("```\n")
        builder.append("[${position + 1}] Time: $time\n")
        builder.append(" - Reason: ${holder.reason}\n")
        builder.append(" - Banned By: ${holder.punisher}\n")
        if (holder.isActive) {
            builder.append(" - Expires: $expiry\n")
        }
        builder.append(" - Active: $active\n")
        builder.append("```")
        return builder
    }

}